File: APP Structure/frontend/src/utils/validators.js
/**
 * Validators
 * Input validation utilities
 */

/**
 * Validate email
 */
export const validateEmail = (email) => {
  const regex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
  return regex.test(email)
}

/**
 * Validate energy type
 */
export const validateEnergyType = (type) => {
  const validTypes = ['solar', 'wind', 'hydro', 'biomass']
  return validTypes.includes(type.toLowerCase())
}

/**
 * Validate quantity
 */
export const validateQuantity = (quantity) => {
  const num = parseFloat(quantity)
  return !isNaN(num) && num > 0
}

/**
 * Validate price
 */
export const validatePrice = (price) => {
  const num = parseFloat(price)
  return !isNaN(num) && num > 0
}

/**
 * Validate order type
 */
export const validateOrderType = (type) => {
  return ['buy', 'sell'].includes(type.toLowerCase())
}

/**
 * Validate required fields
 */
export const validateRequired = (value) => {
  if (typeof value === 'string') {
    return value.trim().length > 0
  }
  return value !== null && value !== undefined
}

/**
 * Validate number range
 */
export const validateRange = (value, min, max) => {
  const num = parseFloat(value)
  return !isNaN(num) && num >= min && num <= max
}

/**
 * Validate trade form
 */
export const validateTradeForm = (formData) => {
  const errors = {}

  if (!validateRequired(formData.energy_type)) {
    errors.energy_type = 'Energy type is required'
  } else if (!validateEnergyType(formData.energy_type)) {
    errors.energy_type = 'Invalid energy type'
  }

  if (!validateRequired(formData.quantity)) {
    errors.quantity = 'Quantity is required'
  } else if (!validateQuantity(formData.quantity)) {
    errors.quantity = 'Quantity must be a positive number'
  }

  if (!validateRequired(formData.price)) {
    errors.price = 'Price is required'
  } else if (!validatePrice(formData.price)) {
    errors.price = 'Price must be a positive number'
  }

  if (!validateRequired(formData.order_type)) {
    errors.order_type = 'Order type is required'
  } else if (!validateOrderType(formData.order_type)) {
    errors.order_type = 'Invalid order type'
  }

  return {
    isValid: Object.keys(errors).length === 0,
    errors
  }
}

File: APP Structure/Backend/src/utils/validators.py
"""
Validators
Input validation utilities
"""

from typing import Any, List


def validate_energy_type(energy_type: str) -> bool:
    """
    Validate energy type
    
    Args:
        energy_type: Energy type to validate
        
    Returns:
        bool: True if valid
    """
    valid_types = ['solar', 'wind', 'hydro', 'biomass']
    return energy_type in valid_types


def validate_price(price: float) -> bool:
    """
    Validate price
    
    Args:
        price: Price to validate
        
    Returns:
        bool: True if valid
    """
    return isinstance(price, (int, float)) and price > 0


def validate_quantity(quantity: float) -> bool:
    """
    Validate quantity
    
    Args:
        quantity: Quantity to validate
        
    Returns:
        bool: True if valid
    """
    return isinstance(quantity, (int, float)) and quantity > 0


def validate_order_type(order_type: str) -> bool:
    """
    Validate order type
    """
    return order_type in ['buy', 'sell']


def validate_email(email: str) -> bool:
    """
    Validate email format
    """
    import re
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return bool(re.match(pattern, email))


def validate_user_id(user_id: str) -> bool:
    """
    Validate user ID format
    """
    return isinstance(user_id, str) and len(user_id) > 0


def validate_contract_status(status: str) -> bool:
    """
    Validate contract status
    """
    valid_statuses = ['pending', 'active', 'completed', 'failed']
    return status in valid_statuses


def validate_transaction_status(status: str) -> bool:
    """
    Validate transaction status
    """
    valid_statuses = ['pending', 'matched', 'completed', 'cancelled']
    return status in valid_statuses


def validate_date_range(start_date: str, end_date: str) -> bool:
    """
    Validate date range
    """
    from datetime import datetime
    
    try:
        start = datetime.fromisoformat(start_date)
        end = datetime.fromisoformat(end_date)
        return start <= end
    except (ValueError, TypeError):
        return False


def validate_required_fields(data: dict, required_fields: List[str]) -> tuple:
    """
    Validate required fields in data
    
    Returns:
        tuple: (is_valid, missing_fields)
    """
    missing = [field for field in required_fields if field not in data]
    return len(missing) == 0, missing
